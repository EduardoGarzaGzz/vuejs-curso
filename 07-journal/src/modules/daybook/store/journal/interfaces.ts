export interface IJournalState {
    isLoading?: boolean,
    entries?: any[]
}
