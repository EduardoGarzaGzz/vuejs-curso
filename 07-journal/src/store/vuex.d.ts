import { IJournalState } from '@/modules/daybook/store/journal/interfaces';
import { Store }         from 'vuex';

declare module '@vue/runtime-core' {

    interface State {
        journal: IJournalState;
    }

    interface ComponentCustomProperties {
        $store: Store<State>;
    }

}
