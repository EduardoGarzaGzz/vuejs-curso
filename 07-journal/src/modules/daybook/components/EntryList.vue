<template>
  <div class="entry-list-container">
    <div class="px-2 pt-2">
      <input v-model="term" class="form-control" placeholder="Buscar entrada" type="text">
    </div>
    <div class="entry-scrollarea">
      <p v-for="entry in entriesByTerm" :key="entry.id"><Entry :entry="entry" /></p>
    </div>
  </div>
</template>

<script lang="ts">
import { defineAsyncComponent } from 'vue';
import { mapGetters }           from 'vuex';

export default {
	name      : 'EntryList',
	components: {
		Entry: defineAsyncComponent( () => import('@/modules/daybook/components/Entry.vue') )
	},
	data() {
		return {
			term: ''
		};
	},
	computed: {
		...mapGetters( 'journal', {
			getEntriesByTerm: 'getEntriesByTerm'
		} ),
		entriesByTerm() {
			return this.getEntriesByTerm( this.term );
		}
	}
};
</script>

<style lang="scss" scoped>
.entry-list-container {
	border-right: 1px solid #2C3E50;
	height: calc(100vh - 56px);
}

.entry-scrollarea {
	height: calc(100vh - 110px);
	overflow: scroll;
}
</style>
