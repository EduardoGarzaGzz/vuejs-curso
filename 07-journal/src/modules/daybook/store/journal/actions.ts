import journalApi        from '@/api/journalApi';
import { IJournalState } from '@/modules/daybook/store/journal/interfaces';
import { Action }        from 'vuex';

export const loadEntries: Action<IJournalState, IJournalState> = async ( { commit } ) => {
    const { data }     = await journalApi.get( `/entries.json` );
    const entries: any = [];

    if ( !data ) {
        commit( 'setEntries', [] );
        return;
    }

    for ( const id of Object.keys( data ) ) {
        entries.push( {
            id,
            ...data[ id ]
        } );
    }

    commit( 'setEntries', entries );
};

export const updateEntry: Action<IJournalState, IJournalState> = async ( { commit }, entry ) => {
    const { id, date, picture, text } = entry;
    const { data }                    = await journalApi.put( `/entries/${ id }.json`, {
        date,
        picture,
        text
    } );
    commit( 'updateEntry', {
        id,
        ...data
    } );
};

export const createEntry: Action<IJournalState, IJournalState> = async ( { commit }, entry ) => {
    const { date, picture, text } = entry;
    const dataToSave              = { date, picture, text };
    const { data }                = await journalApi.post( `/entries.json`, dataToSave );
    commit( 'addEntry', {
        id: data.name,
        ...dataToSave
    } );
    return data.name;
};

export const deleteEntry: Action<IJournalState, IJournalState> = async ( { commit }, entry ) => {
    await journalApi.delete( `/entries/${ entry.id }.json` );
    commit( 'deleteEntry', entry );
};
