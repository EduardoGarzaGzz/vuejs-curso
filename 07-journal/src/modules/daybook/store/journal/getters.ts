import { IJournalState } from '@/modules/daybook/store/journal/interfaces';
import { Getter }        from 'vuex';

export const getEntriesByTerm: Getter<IJournalState, IJournalState> = ( state ) => ( term: string | null = null ): any => {
    const list: any = state.entries;

    if ( !term ) return list;
    return list.filter( ( entry: any ) => entry.text.toLowerCase().includes( term ) );
};


export const getEntriesById: Getter<IJournalState, IJournalState> = ( state ) => ( id = '' ) => {
    const entry = ( state.entries as any ).find( ( entry: any ) => entry.id === id );

    if ( !entry ) return undefined;

    return { ...entry };
};
