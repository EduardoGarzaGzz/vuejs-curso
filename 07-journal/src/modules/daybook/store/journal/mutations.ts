import { IJournalState } from '@/modules/daybook/store/journal/interfaces';
import { Mutation }      from 'vuex';

export const setEntries: Mutation<IJournalState> = ( state, entries: any ) => {
    state.entries   = [ ...( state.entries as any ), ...entries ];
    state.isLoading = false;
};

export const updateEntry: Mutation<IJournalState> = ( state, entryUpdated: any ) => {
    const idx = ( state.entries as any ).map( ( e: any ) => e.id ).indexOf( entryUpdated.id );

    if ( idx !== -1 ) {
        ( state.entries as any )[ idx ] = entryUpdated;
    } else {
        ( state.entries as any ).push( { ...entryUpdated } );
    }
};

export const addEntry: Mutation<IJournalState> = ( state, entry ) => {
    state.entries = [ entry, ...state.entries ];
};

export const deleteEntry: Mutation<IJournalState> = ( state, entry ) => {
    state.entries = state.entries.filter( e => e.id !== entry.id );
};
