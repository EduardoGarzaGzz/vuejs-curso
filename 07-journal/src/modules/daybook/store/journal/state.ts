import { IJournalState } from '@/modules/daybook/store/journal/interfaces';


export const journalState = (): IJournalState => ( {
    isLoading: true,
    entries  : [
        {
            id     : new Date().getTime(),
            date   : new Date().toDateString(),
            text   : 'Danistas accelerare, tanquam barbatus scutum.',
            picture: null
        },
        {
            id     : new Date().getTime() + 1,
            date   : new Date().toDateString(),
            text   : 'Danistas accelerare, tanquam barbatus scutum.',
            picture: null
        },
        {
            id     : new Date().getTime() + 2,
            date   : new Date().toDateString(),
            text   : 'Danistas accelerare, tanquam barbatus scutum.',
            picture: null
        },
        {
            id     : new Date().getTime() + 3,
            date   : new Date().toDateString(),
            text   : 'Danistas accelerare, tanquam barbatus scutum.',
            picture: null
        }
    ]
} );
