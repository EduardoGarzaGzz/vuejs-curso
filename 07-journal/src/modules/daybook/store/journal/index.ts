import * as actions              from '@/modules/daybook/store/journal/actions';
import * as getters              from '@/modules/daybook/store/journal/getters';
import { IJournalState }         from '@/modules/daybook/store/journal/interfaces';
import * as mutations            from '@/modules/daybook/store/journal/mutations';
import { journalState as state } from '@/modules/daybook/store/journal/state';
import { Module }                from 'vuex';

export const journalModule: Module<IJournalState, IJournalState> = {
    namespaced: true,
    state,
    actions,
    mutations,
    getters
};
