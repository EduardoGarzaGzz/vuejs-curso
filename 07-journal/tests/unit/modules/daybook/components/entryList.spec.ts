import EntryList         from '@/modules/daybook/components/EntryList.vue';
import { journalModule } from '@/modules/daybook/store/journal';
import { shallowMount }  from '@vue/test-utils';
import { createStore }   from 'vuex';
import { journalState }  from '../../../mock-data/test-journal-state';

const createVuexStore = ( initialStore: any ) => createStore( {
	modules: {
		journal: {
			...journalModule,
			state: { ...initialStore }
		}
	}
} );

describe( 'Test Component: EntryList.vue', function() {
	const store      = createVuexStore( journalState );
	const mockRouter = {
		push: jest.fn()
	};

	let wrapper: any;

	beforeEach( () => {
		jest.clearAllMocks();

		wrapper = shallowMount( EntryList, {
			global: {
				mocks  : {
					$router: mockRouter
				},
				plugins: [ store ]
			}
		} );
	} );

	it( 'debe de llamar el getEntriesByTerm sin termino y mostrar 2 entradas', async () => {
		expect( wrapper.findAll( 'entry-stub' ).length ).toBe( 1 );
		expect( wrapper.html() ).toMatchSnapshot();
	} );

	it( 'debe de llamar el getEntriesByTerm y filtrar las entradas', () => {
		const input = wrapper.find( 'input' );
		input.setValue( 'Hola' );

		expect( wrapper.findAll( 'entry-stub' ).length ).toBe( 1 );
	} );

	it( 'el buton de nuevo debe de redireccionar a /new', () => {
		const button = wrapper.find( 'button' );
		button.trigger( 'click' );

		expect( mockRouter.push ).toHaveBeenCalledWith( { name: 'entry', params: { id: 'new' } } );
	} );
} );
