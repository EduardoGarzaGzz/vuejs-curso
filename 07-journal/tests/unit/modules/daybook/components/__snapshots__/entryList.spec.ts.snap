// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Component: EntryList.vue debe de llamar el getEntriesByTerm sin termino y mostrar 2 entradas 1`] = `
<div class="entry-list-container">
  <div class="px-2 pt-2"><input class="form-control" placeholder="Buscar entrada" type="text"></div>
  <div class="my-2 d-flex flex-column"><button class="btn btn-primary mx-3"><em class="fa fa-plus-circle"></em> Nueva entrada </button></div>
  <div class="entry-scrollarea">
    <p>
      <entry-stub entry="[object Object]"></entry-stub>
    </p>
  </div>
</div>
`;
